### DISCUSSION TITLE
Report total number of paths vs. number of failing path

### DISCUSSION THREAD

**Discussion by lustefan97** [The-OpenROAD-Project/OpenROAD#4420] [Q&A]
I was wondering if there was a way to quickly summarize the ratio between the number of paths and number of failing paths (with regard to timing) using OpenSTA. I know it can report the slack in each path in a detailled manner, but I would like to have a report in the vein of : "Here is the number of paths that the design has that needs to be tested for timing vs. the number of path that failed the test"

**Comment by maliberty** [comment]
I don't see an easy way as the number of paths is exponential.  How would you use the result?

@tspyrou any thoughts?

**Reply by lustefan97** [reply]
It is just to have a rough idea of how aggressive my constraints are. The WNS is a good indication and a very rough approximation of the number of failing paths can be had by taking the ratio between TNS and WNS if I am not mistaken, but even then I miss the total number of paths (although maybe the reported number of nets would do ?)

**Reply by maliberty** [reply]
The endpoint slack histogram is probably more useful than counting paths.

**Reply by tspyrou** [reply]
Its all in how you define a path. For this need, I think counting each endpoint as a path would probably work well.
The number of failing endpoints / number of endpoints would give a good view of the % of the design which is failing.

**Reply by lustefan97** [reply]
@tspyrou That indeed seems like what I am looking for ! Just to make sure, in my case my designs are mostly just combinatorial circuit whose inputs and outputs are sampled by flip-flops, a path endpoint would be the output FF (in a register-to-register fashion) ? But then how could I make sure that only the slack for those endpoints are reported ?

**Comment by oharboe** [comment]
@lustefan97 Could this be a use-case for timing histogram(supported in OpenROAD under Charts).

**Reply by lustefan97** [reply]
@oharboe If you refer to #2649 , this would be incredible ! But I must admit that after a quick search in OpenROAD's documentation pdf I haven't found anything about charts or how to generate them...

**Reply by oharboe** [reply]
It is there by default, but obviously this feature is effectively impossible to find in the GUI.

It would be better if it was a tab, just like clock tree, timing, hierarchy, etc.

**Reply by lustefan97** [reply]
But then how can we access and create those histograms ?

**Reply by tspyrou** [reply]
@lustefan97 All of the structures needed to create a GUI view are returned by this command.
https://github.com/The-OpenROAD-Project/OpenSTA/blob/42b994d429aef6d553baae6aac8c23477b6e0654/tcl/Search.tcl#L378
set path_ends [find_timing_paths_cmd "report_checks" args]
From this any kind of custom report or display can be developed.
Building a timing histogram view is not in our near terms plans but it is a useful feature.
If your organization was interested in signing up for support you could influence the roadmap.
You could also use the structures in that tcl commmand to generate a report that meets your needs.

**Reply by maliberty** [reply]
@tspyrou we do have a slack histogram already :)

**Reply by oharboe** [reply]
It is just a shame that it is so out if the way, I forget that we do... Can it be moved to the tabs where clocktree, hierarchy, etc. is?

**Comment by maliberty** [comment]
Note that to get the histograms you have to have qt charts installed.   During cmake you should get a message like "Charts widget is enabled" and not "Charts widget is not enabled"

**Reply by oharboe** [reply]
```
sudo apt-get install libqt5charts5-dev
rm ./tools/OpenROAD/build/CMakeCache.txt
./build_openroad.sh --local --latest
```


```
[deleted]
-- Charts widget is enabled
[deleted]
```

**Reply by oharboe** [reply]
This is an endpoint slack histogram, but it looks a bit anaemic for the dataset...

Anyway: I think it would be great if this feature could be made more available out of the box and not hidden away behind "cmake magic" and curious defaults.

If there is only one chart, endpoint slack, then the dropdown should default to endpoint slack. Perhaps with an update button like for timing(since it can take a long time to load timing data).

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/1b7399c8-7b8b-4e1b-b75d-f96125032119)

**Reply by oharboe** [reply]
If I delete the OpenROAD config file, I get:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/9116325a-5ffd-42e8-afef-4cb70e3c56ee)

Whereas, I think this would have been a better, less cramped, default layout:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/3980cf4d-2e0b-4156-91cb-2e8ed7301b59)

**Reply by maliberty** [reply]
A test case for the histogram above would be helpful.

**Reply by oharboe** [reply]
@maliberty I don't think there is a mystery or something wrong... This is just a test module where there are 8 macros and no flip flops. I see ca. 8 endpoints in the histogram, which is correct, I think.

**Comment by maliberty** [comment]
> It is just a shame that it is so out if the way, I forget that we do... Can it be moved to the tabs where clocktree, hierarchy, etc. is?

It is a docking window the same as all the others and can be position wherever you like.

**Reply by oharboe** [reply]
My point is that the default is not good. Why should the default be different for the Charts than e.g. Timing Report?

**Comment by maliberty** [comment]
> Anyway: I think it would be great if this feature could be made more available out of the box and not hidden away behind "cmake magic" and curious defaults.

It is on by default, but qtcharts is not available for centos7 and so can't be assumed.

**Reply by maliberty** [reply]
@vvbandeira we should install qt charts by default in the dependency installer if possible

**Reply by oharboe** [reply]
@vvbandeira @maliberty https://github.com/The-OpenROAD-Project/OpenROAD/issues/4433

**Reply by maliberty** [reply]
@vvbandeira https://github.com/The-OpenROAD-Project/OpenROAD/issues/3523

**Comment by oharboe** [comment]
A less anaemic slack histogram example:

```
make DESIGN_CONFIG=designs/asap7/aes/config.mk cts
```

```
make DESIGN_CONFIG=designs/asap7/aes/config.mk gui_cts
```

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/1caeb16b-836f-4ae5-b5a6-6b235791ab90)

**Reply by lustefan97** [reply]
@oharboe @maliberty Could it be possible to add the WNS on the x-axis to quantify how much slack there is ?

**Reply by maliberty** [reply]
There are labels, eg
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/01b0a15c-6308-43e0-986d-ecaa7ba6ecf5)

but once they get too small they get elided , eg
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/08597006-c348-4b07-a986-9e66359e2b12)

however it is odd that there is no axis label at all in the view from @oharboe 

@AcKoucher please  look at making this more friendly for larger numbers of bin.

**Comment by lustefan97** [comment]
@maliberty @oharboe Ok so I updated my OpenROAD to the latest version while having QT chart installed and enabled. After loading a database in that I knew had reported a slack, I tried to make a timing chart but resulted in the GUI crashing. 

I got a warning telling me that timing data was not stored  in the database and must be loaded separately if needed, are they not stored in the .odb ? Is the .odb file only "geometric" (would not be surprising since a .lef and .def files are needed), even if .lib files and RC rules have been used to produce the final result ?

**Reply by oharboe** [reply]
I get a crash when I load the .odb file and also when I use make gui_floorplan #4435

**Reply by lustefan97** [reply]
How did you obtain the charts in previous posts ?

**Reply by oharboe** [reply]
Follow these instructions: https://github.com/The-OpenROAD-Project/OpenROAD/discussions/4420#discussioncomment-7925602

**Comment by maliberty** [comment]
@AcKoucher FYI

**Comment by oharboe** [comment]
@AcKoucher Just one more example of where the horizontal labels are gone completely:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/7fa91044-d0df-4ffe-84b2-b5a264d959db)

**Reply by lustefan97** [reply]
@AcKoucher @oharboe Note that the "anemic" slack histogram from https://github.com/The-OpenROAD-Project/OpenROAD/discussions/4420#discussioncomment-7921641 also doesn't have any axis, yet is much less populated...

**Reply by AcKoucher** [reply]
@lustefan97 @oharboe I've opened #4447 to fix this.

**Comment by lustefan97** [comment]
@oharboe @maliberty I have a quick but important question, right now I run my TCL scripts on a device that has no GUI, how would I need to proceed if I wanted to design a chip on it using OpenROAD but then move the results and generate the histogram on another computer (capable of opening a GUI).

I ran a test script in which I saved the design in a .odb file and the timing results from OpenSTA in a .sdf file, but when I loaded the .odb back in another GUI session (as well as the .sdf file using `read_sdf`), multiple warnings were raised during loading and no histogram was created...

**Reply by maliberty** [reply]
If you are using ORFS then 'make gui_final' will load everything necessary.

**Reply by lustefan97** [reply]
I do, but my use my own custom scripts... although am I right to assume that putting everything in _flow/results/_ and adjusting the makefile will make things right ?

**Reply by oharboe** [reply]
I use Bazel for this: https://github.com/The-OpenROAD-Project/megaboom

**Reply by maliberty** [reply]
In general you'll need to load the .lib, .odb, .sdc, and .spef
